File: LinkedList/CircularLinkedList/list.py


class Node:
    def __init__(self,data):
        self.data = data
        self.prev = None
        self.next = None

class CircularLinkedList:
    def __init__(self):
        self.head = Node(None)

    def push(self,data):
        newNode = Node(data)
        last = self.head
    
        if(self.head is None):
        	self.head = newNode


        else:
        	while(last.next is not None):
        		last = last.next

        	last.next = newNode
        	newNode.prev = last
        	newNode.next = self.head

    
    def printList(self):
    	tmp = self.head
    	if(self.head is None):
    		print("No list created")

    	while(tmp != self.head):
    		print(self.head.data,'->',end='')
    		tmp = tmp.next


list = CircularLinkedList()

list.push(10)
list.push(11)
list.push(12)
list.push(13)
list.push(14)
list.push(15)


list.printList()

File: LinkedList/DoublyLinkedList/List.c

#include <stdio.h>

struct Node
{
    int data;
    struct Node* next;
    struct Node* prev;
};

void pushFront(struct Node** head,int data){

    struct Node* new_node = (struct Node*)malloc(sizeof(struct Node)); 
 
    new_node->data = data; 
 
    new_node->next = *head; 

    new_node->prev = NULL; 
 
    if ((*head) != NULL) 
        (*head)->prev = new_node; 
 
    *head = new_node; 
};

void append(struct Node** head,int data){
    struct Node* new_node = (struct Node*)malloc(sizeof(struct Node)); 

    new_node->data = data;

    struct Node* last = *head;
    
    new_node->next = NULL;

    if (*head == NULL)
    { 
        new_node->prev = NULL; 
        *head = new_node; 
        return; 
    } 

    while (last->next != NULL) 
        last = last->next; 
 
    last->next = new_node; 
    new_node->prev = last; 
};

void printList(struct Node* head) 
{ 
    struct Node* last; 

    printf("\nTraversal in forward direction \n"); 
    while (head != NULL) { 
        printf(" %d ", head->data); 
        last = head; 
        head = head->next; 
    } 
 
    printf("\nTraversal in reverse direction \n"); 
    while (last != NULL) { 
        printf(" %d ", last->data); 
        last = last->prev; 
    } 
}; 

void insertBefore(struct Node** head , int position , int data){

        struct Node* pos;
        struct Node* temp = head;

        struct Node* new_node = (struct Node*)malloc(sizeof(struct Node)); 

        if(head == NULL){
            return;
        };

        for(int i = 0; i < position;i++){
            temp = temp->next;
        };    

        new_node->data = data;

        new_node->prev = temp->prev;
        temp->prev = new_node;
        new_node->next = temp;                  

         if (new_node->prev != NULL) 
            new_node->prev->next = new_node; 

        else
            head = new_node;        
};

void main() {

    struct Node* head = NULL; 

    append(&head, 1);      
    append(&head, 2);  
    append(&head, 3); 
    append(&head, 4);      
    append(&head, 5);  
    append(&head, 6); 
 
    printf("Created DLL is: "); 
    printList(head);     

    insertBefore(head,3,11);

    printf("List after new insertion");

    printList(head);
}
